Add procedure on configuring HPA for an app that exposes metrics

rh-max opened the pull request.

vikram-redhat commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+[id='configuring-hpa-for-an-application']
```
Double Quotes

vikram-redhat commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+//
+// * machine_management/configuring-hpa-for-an-application.adoc
+
+[id='configuring-hpa-based-on-application-metrics-{context}']
```
Double Quotes

kalexand-rh commented on _topic_map.yml:
```diff
@@ -335,6 +335,8 @@ Topics:
   File: creating-machineset
 - Name: Creating infrastructure MachineSets
   File: creating-infrastructure-machinesets
+- Name: Configuring Horizontal Pod Autoscaling for an application
```
This doesn't go in *Machine management*. I think it might go in *Applications*.

kalexand-rh commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+
+.Prerequisites
+
+* Log in as a user with administrative privileges for the cluster.
```
```
* Install the OpenShift Command-line Interface (CLI), commonly known as `oc`.
* You must log in to the cluster with a user that has the `cluster-admin` role.
```

kalexand-rh commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+.Prerequisites
+
+* Log in as a user with administrative privileges for the cluster.
+* Make sure you configured the cluster for application monitoring. In the following example, it is presumed that Prometheus and Alertmanager instances were installed in the `default` namespace. See xref:../monitoring/application-monitoring.adoc#application-monitoring[Application monitoring].
```
> Make sure you configured the cluster for application monitoring.

Add a link in the sentence to the correct assembly for configuring the cluster for monitoring.

kalexand-rh replied:
What do you do if they're not in the same namespace? Is making them be in the same namespace a reasonable assumption?

kalexand-rh commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+
+* Log in as a user with administrative privileges for the cluster.
+* Make sure you configured the cluster for application monitoring. In the following example, it is presumed that Prometheus and Alertmanager instances were installed in the `default` namespace. See xref:../monitoring/application-monitoring.adoc#application-monitoring[Application monitoring].
+* Make sure you configured monitoring for your application. In the following example, it is presumed that the application and the service monitor for it were installed in the `default` namespace. See xref:../monitoring/application-monitoring.adoc#application-monitoring[Application monitoring].
```
> Make sure that you configured monitoring for your application.

Add a link to the right section of the docs to configure monitoring for the application.  Consider including the modules instead of the prereq link.

kalexand-rh commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+* Log in as a user with administrative privileges for the cluster.
+* Make sure you configured the cluster for application monitoring. In the following example, it is presumed that Prometheus and Alertmanager instances were installed in the `default` namespace. See xref:../monitoring/application-monitoring.adoc#application-monitoring[Application monitoring].
+* Make sure you configured monitoring for your application. In the following example, it is presumed that the application and the service monitor for it were installed in the `default` namespace. See xref:../monitoring/application-monitoring.adoc#application-monitoring[Application monitoring].
+* Make sure your application exposes metrics using the `prometheus-adapter`. In the following example, it is presumed that `prometheus-adapater` was installed in the `default` namespace. See xref:../monitoring/application-monitoring.adoc#application-monitoring[Application monitoring].
```
> Make sure that your application exposes metrics using the `prometheus-adapter`

Add a link to the right section of the docs to expose the metrics. Consider including the modules instead of the prereq link.

kalexand-rh commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+
+include::modules/configuring-hpa-based-on-application-metrics.adoc[leveloffset=+1]
+
+.Additional resources
```
I don't think you need this. You've already linked to the monitoring sections three times.

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+[id='configuring-hpa-based-on-application-metrics-{context}']
+= Configuring HPA based on application metrics
+
+If you have configured an application to export metrics, you can set up Horizontal Pod Autoscaling (HPA) based on these metrics.
```
s/have configured/configure

kalexand-rh commented on machine_management/configuring-hpa-for-an-application.adoc:
```diff
@@ -0,0 +1,25 @@
+
+toc::[]
+
+You can configure Horizontal Pod Autoscaling (HPA) for an application that exports metrics.
```
What is HPA? I think you need a conceptual module about it. You need to explain what it is, when you might use it, and what happens if you use it and cluster autoscaling.

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+
+.Procedure
+
+. Create a YAML file for your configuration. In this example, it is called `deploy.yaml`.
```
You can merge the first two steps.

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+
+. Create a YAML file for your configuration. In this example, it is called `deploy.yaml`.
+
+. Add configuration for deploying the horizontal pod autoscaler for the application. This example configures and deploys HPA based on the application `http_requests_per_second` metric for the sample application configured in the "Application monitoring" section:
```
Remove the second sentence and replace it with annotations in the yaml.

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+<4> `metricName` specifies the metric upon which HPA is based
+<5> `targetAverageValue` specifies the value of the metric that HPA should try to maintain
+
+. Apply the configuration file to the cluster:
```
You must enclose the command in
```
----
----
```

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+      metricName: http_requests_per_second <4>
+      targetAverageValue: 10 <5>
+----
+<1> `name` specifies the application
```
End each annotation with a period.

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+      targetAverageValue: 10 <5>
+----
+<1> `name` specifies the application
+<2> `minReplicas` specifies the minimum number of replicas the HPA should maintain for the application
```
s/should maintain/maintains

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+----
+<1> `name` specifies the application
+<2> `minReplicas` specifies the minimum number of replicas the HPA should maintain for the application
+<3> `maxReplicas` specifies the maximum number of replicas the HPA should maintain for the application
```
s/should maintain/maintains

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+<2> `minReplicas` specifies the minimum number of replicas the HPA should maintain for the application
+<3> `maxReplicas` specifies the maximum number of replicas the HPA should maintain for the application
+<4> `metricName` specifies the metric upon which HPA is based
+<5> `targetAverageValue` specifies the value of the metric that HPA should try to maintain
```
"should try to maintain" - be more specific. Don't use "should."

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+<1> `name` specifies the application
+<2> `minReplicas` specifies the minimum number of replicas the HPA should maintain for the application
+<3> `maxReplicas` specifies the maximum number of replicas the HPA should maintain for the application
+<4> `metricName` specifies the metric upon which HPA is based
```
Can you use any valid metric in the cluster, or are there limits? Please clarify.

kalexand-rh commented on modules/configuring-hpa-based-on-application-metrics.adoc:
```diff
@@ -0,0 +1,44 @@
+
+. Apply the configuration file to the cluster:
++
+  $ oc apply -f deploy.yaml
```
How do I know this worked?

kalexand-rh commented on monitoring/application-monitoring.adoc:
```diff
@@ -14,3 +14,7 @@ include::modules/monitoring-configuring-cluster-for-application-monitoring.adoc[
 
+.Next steps
+
+* xref:../monitoring/application-monitoring.adoc#application-monitoring[Configure Horizontal Pod Autoscaling for the application.]
```
You need to provide more context about why you'd want to do this.

rh-max commented:
@kalexand-rh Thank you for the review. Most feedback is implemented, and some items are awaiting info from the team.

One update on this: the team has decided to not include the application monitoring docs for this release (they will be in a later one). So I've removed the section that this PR introduces from _topic_map.yml and moved the files to temporarily stay in monitoring/. Just FYI.

vikram-redhat commented:
@rh-max - please make sure to merge this by Tuesday COB, your time. If QE requests updates on this, we can look at that as appropriate.
